src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+
+            this.CheckMaxReorgRuleViolated(newChainedHeader);
+
+            previousChainedHeader = newChainedHeader;
+
+            for (; newHeaderPosition < headers.Count; newHeaderPosition++)
+            {
+                newChainedHeader = this.CreateAndValidateNewChainedHeader(newChainedHeaders, headers[newHeaderPosition], previousChainedHeader);
+                this.logger.LogTrace("New chained header was added to the tree '{0}'.", newChainedHeader);
+
+                previousChainedHeader = newChainedHeader;
+            }
+
+            this.logger.LogTrace("(-):*.{0}:{1}", nameof(newChainedHeaders.Count), newChainedHeaders.Count);
+            return newChainedHeaders;
+        }
+
+        private ChainedHeader CreateAndValidateNewChainedHeader(List<ChainedHeader> newChainedHeaders, BlockHeader currentBlockHeader, ChainedHeader previousChainedHeader)
+        {
+            var newChainedHeader = new ChainedHeader(currentBlockHeader, currentBlockHeader.GetHash(), previousChainedHeader);
+
+            this.chainedHeaderValidator.Validate(newChainedHeader);
+
+            newChainedHeaders.Add(newChainedHeader);
+            previousChainedHeader.Next.Add(newChainedHeader);
+            this.chainedHeadersByHash.Add(newChainedHeader.HashBlock, newChainedHeader);
+
+            return newChainedHeader;
+        }
+
+        private bool FindFirstNewHeader(List<BlockHeader> headers, out int newHeaderPosition)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(headers), nameof(headers.Count), headers.Count);
+
+            for (newHeaderPosition = 0; newHeaderPosition < headers.Count; newHeaderPosition++)
+            {
+                uint256 currentBlockHash = headers[newHeaderPosition].GetHash();
+                if (!this.chainedHeadersByHash.ContainsKey(currentBlockHash))
+                {
+                    this.logger.LogTrace("A new header with hash '{0}' was found that is not connected to the tree.", currentBlockHash);
+                    this.logger.LogTrace("(-):true");
+                    return true;
+                }
+            }
+
+            this.logger.LogTrace("(-):false,{0}:{1}", nameof(newHeaderPosition), newHeaderPosition);
+            return false;
+        }
+
+        /// <summary>
+        /// Checks if <paramref name="chainedHeader"/> violates the max reorg rule, if networks.Consensus.MaxReorgLength is zero this logic is disabled.
+        /// </summary>
+        /// <param name="chainedHeader">The header that needs to be checked for reorg.</param>
+        /// <returns><c>true</c> if maximum reorg rule violated, <c>false</c> otherwise.</returns>
+        private void CheckMaxReorgRuleViolated(ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            ChainedHeader tip = this.chainState.ConsensusTip.FindFork(chainedHeader);
```

Comment:
This can be done after `if ((maxReorgLength != 0) && (consensusTip != null))`

Reply:
Also why fork point is being found 2 times? 

`tip` and `fork` will be equal here

probably the whole line 
`ChainedHeader tip = this.chainState.ConsensusTip.FindFork(chainedHeader);`
is redundant.